### DISCUSSION TITLE
Default value for category field

### DISCUSSION THREAD

**Discussion by Century1026** [inventree/InvenTree#7259] [Q&A]
Hello,

As you can see, name field can display the first item on category list with default value, but when I add `default: PartCategory.objects.first()` to category field, the form isn't even opening.

How can I add a default value for category field?
Looking forward to your answer!

partial code in part.js
```
function partFields(options={}) {

    var fields = {
        category: {
            filters: {
                structural: false,
            },
            tree_picker: {
                url: '{% url "api-part-category-tree" %}',
                default_icon: global_settings.PART_CATEGORY_DEFAULT_ICON,
            },
        },
        name: {
            default: PartCategory.objects.first().name,
        },
    if (options.category) {
        fields.category.value = options.category;
    }
    return fields;
}
```

**Comment by SchrodingersGat** [comment]
You cannot use `PartCategory.objects.first().name` in front-end javascript code. This is a database call, made by python / django in the backend.

**Reply by Century1026** [reply]
Hi, Oliver
Thank you for your prompt answer!
I'm sorry I get it wrong, `PartCategory.objects.first().name` works here in models.py.
```
    name = models.CharField(
        max_length=100, 
        blank=False, 
        default=PartCategory.objects.first().name,
        help_text=_('Part name'), 
        verbose_name=_('Name')
    )
```
But why adding `default=PartCategory.objects.first()` is not working for category? Does it require id or pk number? If so, how can I fetch it?
models.py
```
category = TreeForeignKey(
    PartCategory,
    related_name='parts',
    null=True,
    blank=True,
    on_delete=models.DO_NOTHING,
    default=PartCategory.objects.first(),#nothing happened?
    verbose_name=_('Category'),
    help_text=_('Part category'),
)
```

**Reply by Century1026** [reply]
A little bit of context here...

I noticed that user can intitialize stock quantity and location when creating part. It's efficient as additional steps to edit stock are saved. As expiry date is also a primary prop, I added "Initial Stock Expiry Date" field to "Create Part" form and it worked perfectly(see attached video).
https://github.com/inventree/InvenTree/assets/106154819/1c2d4cc6-29d7-419b-a68c-9ab3d6d84186



To further improve efficiency, I wanna realize that when "initial location" is selected, the name field instantly generate a ref number like A-a-1, with each character represents locations and sublocations.

Therefore, to save time creating testing parts, I hope fields can be auto-filled. Unnecessary fields are hid and initial stock quantity field is able to generate a random number (that's why I was making changes in part.js). Currently I only have category field un-auto-filled.

**Reply by SchrodingersGat** [reply]
Default field values cannot be specified "dynamically" like that - you should read more about [how field defaults work](https://docs.djangoproject.com/en/4.2/ref/models/fields/)

The "default" value gets evaluated only once, when the code first runs - and *not* every time you instantiate a field.

The way to do it dynamically is to inject a new default value into the form on the frontend. The new (React) interface has a much more comprehensive and powerful framework for achieving this.

**Reply by Century1026** [reply]
Thank you for your kind guidance, I'm getting more familiar of how it works!

**Comment by Century1026** [comment]
I suddenly realize that when editing a part, the category and name field are auto-filled. Maybe I could learn the way it presents value just like how I added initial_stock_expiry_date.

In the latter part of part.js, I found this
```
function editPart(pk) {

    var url = `{% url "api-part-list" %}${pk}/`;

    var fields = partFields({
        edit: true
    });

    // Filter supplied parts by the Part ID
    fields.default_supplier.filters.part = pk;

    var groups = partGroups({});

    constructForm(url, {
        fields: fields,
        groups: groups,
        title: '{% trans "Edit Part" %}',
        reload: true,
        successMessage: '{% trans "Part edited" %}',
    });
}
```

The function accepts a pk number to fetch value, I then searched for "editpart" and found this
```
    $("#part-edit").click(function() {

        editPart({{ part.pk }});
    });
    {% endif %}
```
Apparently, when you click editpart button, you are already under certain part info page. But when creating new part, no pk can be fetched.

**Comment by wolflu05** [comment]
Regarding the python (backend) and javascript (frontend) mixes, refer to the comments above. But if you want to build a system that automatically sets fields for you, I would suggest creating a [custom plugin](https://docs.inventree.org/en/stable/extend/plugins/) rather than running a modified version of InvenTree. There are several mixing available. I would recommend taking a look at the [event mixin](https://docs.inventree.org/en/stable/extend/plugins/event/) which you could use to set certain values on a item creation.

**Reply by Century1026** [reply]
Thank you for your answer! Will look into it.
